File: api/app/services/site_config.py
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.site_settings import SiteSetting
from app.schemas.site_config import HomepageSection, PricingConfig, PromoTile, SiteConfig, ThemeConfig

SITE_CONFIG_KEY = "homepage"

DEFAULT_SECTIONS: list[HomepageSection] = [
    HomepageSection(id="hero", type="hero", enabled=True, order=0),
    HomepageSection(id="guarantee", type="guarantee", enabled=True, order=1),
    HomepageSection(id="mega_categories", type="mega_categories", enabled=True, order=2),
    HomepageSection(id="brands", type="brands", enabled=True, order=3),
    HomepageSection(id="gender_tabs", type="gender_tabs", enabled=True, order=4),
    HomepageSection(id="footwear", type="footwear", enabled=True, order=5),
    HomepageSection(id="clothing", type="clothing", enabled=True, order=6),
    HomepageSection(id="lookbook", type="lookbook", enabled=False, order=7),
    HomepageSection(id="accessories", type="accessories", enabled=True, order=8),
    HomepageSection(id="curated", type="curated", enabled=False, order=9),
    HomepageSection(id="pickup", type="pickup", enabled=False, order=10),
    HomepageSection(id="gift_cert", type="gift_cert", enabled=False, order=11),
]

DEFAULT_PROMO_TILES: list[PromoTile] = [
    PromoTile(
        id="garage_sale",
        label="Garage Sale",
        sublabel="С 24 по 31 августа",
        color="#ff385c",
        href="/catalog",
        enabled=True,
        order=0,
    ),
    PromoTile(
        id="discount",
        label="Забери скидку",
        sublabel="до 31 августа",
        color="#7c3aed",
        href="/catalog",
        enabled=True,
        order=1,
    ),
    PromoTile(
        id="app",
        label="Скачай приложение",
        sublabel="Получи скидку",
        color="#222222",
        href="/catalog",
        enabled=False,
        order=2,
    ),
    PromoTile(
        id="photo_search",
        label="Найди товар по фотке",
        sublabel="Байер сервис",
        color="#222222",
        href="/search",
        enabled=False,
        order=3,
    ),
]


def default_site_config() -> SiteConfig:
    return SiteConfig(
        theme=ThemeConfig(),
        sections=DEFAULT_SECTIONS,
        tile_overrides=[],
        promo_tiles=DEFAULT_PROMO_TILES,
        pricing=PricingConfig(),
        remove_bg_default=True,
    )


async def get_site_config(session: AsyncSession) -> SiteConfig:
    row = await session.get(SiteSetting, SITE_CONFIG_KEY)
    if not row:
        return default_site_config()
    data = row.value
    base = default_site_config().model_dump()
    base.update(data)
    if not base.get("promo_tiles"):
        base["promo_tiles"] = [p.model_dump() for p in DEFAULT_PROMO_TILES]
    if not base.get("pricing"):
        base["pricing"] = PricingConfig().model_dump()
    return SiteConfig.model_validate(base)


async def save_site_config(session: AsyncSession, config: SiteConfig) -> SiteConfig:
    row = await session.get(SiteSetting, SITE_CONFIG_KEY)
    payload = config.model_dump()
    if row:
        row.value = payload
    else:
        session.add(SiteSetting(key=SITE_CONFIG_KEY, value=payload))
    await session.commit()
    return config

File: web/src/lib/site-config.ts
import { apiFetch } from "@/lib/api";

export type ThemeConfig = {
  accent: string;
  accent_dark: string;
  accent_light: string;
  cart_button: string;
  hero_yellow: string;
};

export type HomepageSection = {
  id: string;
  type: string;
  enabled: boolean;
  order: number;
  title?: string | null;
};

export type TileOverride = {
  id: string;
  label: string;
  image_url?: string | null;
  category_slug?: string | null;
  keywords?: string[];
};

export type PromoTile = {
  id: string;
  label: string;
  sublabel?: string | null;
  color: string;
  href: string;
  enabled: boolean;
  order: number;
};

export type PricingConfig = {
  default_markup_percent: number;
  site_discount_percent: number;
};

export type SiteConfig = {
  theme: ThemeConfig;
  sections: HomepageSection[];
  tile_overrides: TileOverride[];
  promo_tiles: PromoTile[];
  pricing: PricingConfig;
  remove_bg_default: boolean;
};

export const DEFAULT_PROMO_TILES: PromoTile[] = [
  {
    id: "garage_sale",
    label: "Garage Sale",
    sublabel: "С 24 по 31 августа",
    color: "#ff385c",
    href: "/catalog",
    enabled: true,
    order: 0,
  },
  {
    id: "discount",
    label: "Забери скидку",
    sublabel: "до 31 августа",
    color: "#7c3aed",
    href: "/catalog",
    enabled: true,
    order: 1,
  },
  {
    id: "app",
    label: "Скачай приложение",
    sublabel: "Получи скидку",
    color: "#222222",
    href: "/catalog",
    enabled: false,
    order: 2,
  },
  {
    id: "photo_search",
    label: "Найди товар по фотке",
    sublabel: "Байер сервис",
    color: "#222222",
    href: "/search",
    enabled: false,
    order: 3,
  },
];

export const DEFAULT_SECTIONS: HomepageSection[] = [
  { id: "hero", type: "hero", enabled: true, order: 0 },
  { id: "guarantee", type: "guarantee", enabled: true, order: 1 },
  { id: "mega_categories", type: "mega_categories", enabled: true, order: 2 },
  { id: "brands", type: "brands", enabled: true, order: 3 },
  { id: "gender_tabs", type: "gender_tabs", enabled: true, order: 4 },
  { id: "footwear", type: "footwear", enabled: true, order: 5 },
  { id: "clothing", type: "clothing", enabled: true, order: 6 },
  { id: "lookbook", type: "lookbook", enabled: false, order: 7 },
  { id: "accessories", type: "accessories", enabled: true, order: 8 },
  { id: "curated", type: "curated", enabled: false, order: 9 },
  { id: "pickup", type: "pickup", enabled: false, order: 10 },
  { id: "gift_cert", type: "gift_cert", enabled: false, order: 11 },
];

export const DEFAULT_SITE_CONFIG: SiteConfig = {
  theme: {
    accent: "#7C3AED",
    accent_dark: "#5B21B6",
    accent_light: "#EDE9FE",
    cart_button: "#7C3AED",
    hero_yellow: "#EDE9FE",
  },
  sections: DEFAULT_SECTIONS,
  tile_overrides: [],
  promo_tiles: DEFAULT_PROMO_TILES,
  pricing: { default_markup_percent: 0, site_discount_percent: 0 },
  remove_bg_default: true,
};

export const SECTION_LABELS: Record<string, string> = {
  hero: "Главный баннер",
  guarantee: "Гарантия",
  mega_categories: "Категории (горизонт)",
  brands: "Бренды",
  gender_tabs: "Пол",
  footwear: "Обувь",
  clothing: "Одежда",
  lookbook: "Лукбук",
  accessories: "Аксессуары",
  curated: "Подборки",
  pickup: "Забрать сегодня",
  gift_cert: "Подарочный сертификат",
};

export async function getSiteConfig(): Promise<SiteConfig> {
  try {
    const data = await apiFetch<SiteConfig>("/site/config", { next: { revalidate: 30 } });
    return {
      ...DEFAULT_SITE_CONFIG,
      ...data,
      theme: { ...DEFAULT_SITE_CONFIG.theme, ...data.theme },
      promo_tiles: data.promo_tiles?.length ? data.promo_tiles : DEFAULT_PROMO_TILES,
      pricing: data.pricing ?? DEFAULT_SITE_CONFIG.pricing,
    };
  } catch {
    return DEFAULT_SITE_CONFIG;
  }
}

export async function getCategoryThumbnails(): Promise<Record<string, string>> {
  try {
    return await apiFetch<Record<string, string>>("/catalog/category-thumbnails", {
      next: { revalidate: 120 },
    });
  } catch {
    return {};
  }
}

export function sortedSections(config: SiteConfig): HomepageSection[] {
  return [...config.sections]
    .filter((s) => s.enabled)
    .sort((a, b) => a.order - b.order);
}

export function themeCssVars(theme: ThemeConfig): Record<string, string> {
  return {
    "--color-accent": theme.accent,
    "--color-accent-dark": theme.accent_dark,
    "--color-accent-light": theme.accent_light,
    "--color-cart-pink": theme.cart_button,
    "--color-hero-yellow": theme.hero_yellow,
    "--color-navy": theme.accent_dark,
  };
}
